rust-runtime/aws-smithy-http-server/src/routing/request_spec.rs

```diff
@@ -118,50 +111,102 @@ impl From<&PathSpec> for Regex {
     }
 }
 
-impl RequestSpec {
+/// Trait definition of a request specification, used to define different routing strategies.
+pub trait RequestSpec {
+    /// Validate the route against the current request and return a [`Match`] enum.
+    fn matches<B>(&self, req: &Request<B>) -> Match;
+
+    /// This method is used to rank how "important" is a route as some protocols defined by Smithy
+    /// (Ie RestJson1 or RestXml) can have conflicting routes.
+    ///
+    /// Other protocols (Ie AwsJson 1.0 and 1.1) cannot have conflicting routes since every
+    /// request MUST be sent to the root URL (/), hence the default implementation returns 0.
+    fn rank(&self) -> usize {
+        0
+    }
+}
+
+/// Request specification for AwsJson 1.0 and 1.1.
+/// Both protocols routing is based on the header `x-amz-target`, which holds the selected operation.
+#[derive(Debug, Clone)]
+pub struct AwsJsonRequestSpec {
+    x_amzn_target: String,
+}
+
+impl AwsJsonRequestSpec {
+    /// Create a new [`AwsJsonRequestSpec`]
+    pub fn new(x_amzn_target: String) -> Self {
+        Self { x_amzn_target }
+    }
+}
+
+impl RequestSpec for AwsJsonRequestSpec {
+    fn matches<B>(&self, req: &Request<B>) -> Match {
+        // We first check for the method and URI of the current HTTP request.
+        // Every request for the AwsJson 1.0 and 1.1 protocol MUST be sent to the root URL (/)
+        // using the HTTP "POST" method.
+        if req.method() != http::Method::POST {
```

Comment:
`MethodNotAllowed` should be sent when the resource (i.e. URI) is correct, but the method is not, so this check should go below checking for `/`.

Add a unit test to not regress on this.

Reply:
Good catch!